File: algorithms/653. Two Sum IV - Input is a BST/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findTarget(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: bool
        """
        counter = collections.defaultdict(int)
        def dfs(node):
            if node:
                counter[node.val] += 1
                dfs(node.left)
                dfs(node.right)
        dfs(root)
        for val in counter:
            if k - val == val:
                if counter[val] > 1:
                    return True
            elif k - val in counter:
                return True
        return False

# 421 / 421 test cases passed.
# Status: Accepted
# Runtime: 198 ms

File: algorithms/653. Two Sum IV - Input is a BST/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        unordered_map<int, int> counter;
        dfs(counter, root);
        for ( auto & p : counter ) {
            if ( 2 * p.first == k ) {
                if ( p.second > 1 )
                    return true;
            } else if ( counter.count(k - p.first) )
                return true;
        }
        return false;
    }

    void dfs(unordered_map<int, int> & counter, TreeNode * root) {
        if ( root ) {
            ++counter[root->val];
            dfs(counter, root->left);
            dfs(counter, root->right);
        }
    }
};

// 421 / 421 test cases passed.
// Status: Accepted
// Runtime: 63 ms
